lib/vsc/filesystem/oceanstor.py

```diff
@@ -405,3 +416,153 @@ def list_filesets(self, devices=None, filesystemnames=None, filesetnames=None, u
+        del parent_fileset_name
+        del afm
+        del inodes_max
+        del inodes_prealloc
+
+        dt_fullpath = self._sanity_check(new_fileset_path)
+
+        if fileset_name is None:
+            # Use name of last folder as name of dtree fileset in OceanStor
+            fileset_name = os.path.basename(dt_fullpath)
+        else:
+            # Append the fileset name to the given path
+            dt_fullpath = os.path.join(dt_fullpath, fileset_name)
+            dt_fullpath = self._sanity_check(dt_fullpath)
+
+        # Check existence of path in local filesystem
+        if self.exists(dt_fullpath):
+            errmsg = "Path of new fileset '%s' is validated as '%s' but it already exists."
+            self.log.raiseException(errmsg % (new_fileset_path, dt_fullpath), OceanStorOperationError)
+
+        dt_parentdir = os.path.dirname(dt_fullpath)
+        if not self.exists(dt_parentdir):
+            errmsg = "Parent directory '%s' of new fileset '%s' does not exist. It will not be created automatically."
+            self.log.raiseException(errmsg % (dt_parentdir, dt_fullpath), OceanStorOperationError)
+
+        # Get local mounted filesystem and remote one in OceanStor
+        local_fs = self.what_filesystem(dt_parentdir)
+        local_mount = local_fs[self.localfilesystemnaming.index('mountpoint')]
+        oceanstor_fs_name = local_fs[self.localfilesystemnaming.index('oceanstorfs')]
+
+        # Relative path of parent directory holding the fileset
+        dt_parentdir_relative = os.path.relpath(dt_parentdir, start=local_mount)
+        if dt_parentdir_relative.startswith('..'):
+            errmsg = "Parent directory of new fileset cannot be outside of filesystem boundaries. %s got: '%s'"
+            self.log.raiseException(errmsg % (fileset_name, dt_parentdir_relative), OceanStorOperationError)
+        elif dt_parentdir_relative == '.':
+            dt_parentdir_relative = ''
+
+        oceanstor_parentdir = '/' + dt_parentdir_relative
+        self.log.debug("Creating new dtree fileset with parent directory: '%s'", oceanstor_parentdir)
+
+        # Send API request to create the new dtree fileset
+        self.make_fileset_api(fileset_name, oceanstor_fs_name, parent_dir=oceanstor_parentdir)
+
+    def make_fileset_api(self, fileset_name, filesystem_name, parent_dir='/'):
+        """
+        Create new dtree fileset in given filesystem of OceanStor
+        
+        - Dtree filesets cannot be nested
+        - Dtree filesets can be created at specific path inside the filesystem,
+          but this information cannot be retrieved back from the OceanStor API
+          (list_filesets()). Therefore, all filesets in a common filesystem must
+          have unique names.
+
+        @type fileset_name: string with the name of the new fileset
+        @type filesystem_name: string with the name of an existing filesystem
+        @type filesystem_name: string with the path of parent directory of new fileset
```

Comment:
this is double?

Reply:
fixed in https://github.com/vub-hpc/vsc-filesystem-oceanstor/pull/3/commits/c3bfb89c5e1696a43e4f0d481e9a90785d0ab3c6